src/ray/core_worker/context.cc

```diff
@@ -303,15 +314,23 @@ void WorkerContext::SetCurrentTask(const TaskSpecification &task_spec) {
   } else {
     RAY_CHECK(false);
   }
+
   if (task_spec.IsNormalTask() || task_spec.IsActorCreationTask()) {
-    // TODO(architkulkarni): Once workers are cached by runtime env, we should
-    // only set runtime_env_ once and then RAY_CHECK that we
-    // never see a new one.
-    runtime_env_info_ = std::make_shared<rpc::RuntimeEnvInfo>();
-    *runtime_env_info_ = task_spec.RuntimeEnvInfo();
-    if (!IsRuntimeEnvEmpty(runtime_env_info_->serialized_runtime_env())) {
-      runtime_env_ = std::make_shared<nlohmann::json>();
-      *runtime_env_ = nlohmann::json::parse(runtime_env_info_->serialized_runtime_env());
+    RAY_CHECK(runtime_env_info != nullptr);
+    runtime_env_info_ = std::move(runtime_env_info);
+
+    if (!IsRuntimeEnvEmpty(serialized_runtime_env)) {
+      // All threads are requesting for the same parsed json result, so ok to place in
+      // critical section.
+      if (!serialized_runtime_env.empty()) {
+        RAY_CHECK_EQ(cached_serialized_runtime_env_, serialized_runtime_env)
```

Comment:
cached_serialized_runtime_env_ can be empty if it's the first time?

Reply:
it's a typo... I should use `serialized_runtime_env_.empty()`; fixed